### ISSUE TITLE
whisper demo recognize wrong text (some segment text repeat again and again)

### ISSUE THREAD

**Issue by makao007** [huggingface/candle#472]
I tried to build and run like below, and found the segment text recognize wrong.
```
cargo run --example whisper --release
cd target/release/examples
whisper.exe --input 001.wav --model base-en
```
many same segment text repeat again and again.
but the OpenAI/Whisper Pytorch Version can give me the right text.

my wav audio file 
https://drive.google.com/file/d/1qwQwfDK-rzac2mnAlSE-oXsKzlKoiA74/view?usp=sharing

**Comment by LaurentMazare** [comment]
Thanks for reporting this, I tried it out and observed the same thing: the candle version repeats some bits a large number of times whereas the python based version seems to decode things fine. It seems to happen with the default model (tiny.en) but also with larger models though differently.
We'll investigate and try to make things line up between candle and the python version but this may take a bit of time.

**Comment by Narsil** [comment]
Whisper runs in `timestamp tokens` mode by default: https://github.com/openai/whisper/blob/main/whisper/decoding.py#L108-L109

This adds this logits processor: https://github.com/openai/whisper/blob/main/whisper/decoding.py#L554-L565

I suspect all models where trained with this one, because the hallucination does happen much more when not using it.
Also whisper is quite sensitive to initial audio conditions. If you cut audio files brutally, it will tend to hallucinate much more. It works a bit better if you cut the audio at the `timestamp_token` boundaries that it emits in that mode.

I think the fallback is already implemented.

It can also be a more classical bug.

**Comment by LaurentMazare** [comment]
Yeah I'm actually making some changes to reflect the initial tokens including the timestamp bits more accurately and this seems to improve things significantly though it's hard to replicate the whole whisper logic as it's rather convoluted.

**Comment by LaurentMazare** [comment]
I just merge #491 as a quick workaround, it's better on some sequences including this sample one, but slightly worse on some others (though it's also the case for the openai implementation). We will work on proper timestamp support but hopefully this is good enough for your use cases at the moment (if you notice that not enough text is recognized, you can pass `--task transcribe` on top it, for some reason the openai implementation doesn't always include this transcribe token in the start tokens so I've just replicated their logic here and you have to ask for it explicitely - I'll test more and this might become the default if it appears to be better in all the cases that I see).

**Comment by LaurentMazare** [comment]
Closing this for now as hopefully it helped with your example, feel free to re-open or create new issues if you come along examples where the decoding is still flaky.

**Comment by LaurentMazare** [comment]
Just to follow up on this, we've merged #539 which improves the timestamp support, e.g. if you run on your example with the `--timestamps` argument you should get the following (it's far from perfect but that's running with the tiny.en model).
```
./whisper --input001.wav --timestamps`
0.0s -- 30.0s
  0.0s-29.0s:  Hello English learners and welcome to English Pod. My name is Marco. I'm Amira and Amira and I are here today with a great great lesson for you. Yes, we are. Today we're going to be talking about a restaurant. Amira, why don't you give us a little bit more details? Well, we're talking about a situation in a restaurant and two people are involved, the waiter and the customer.
30.0s -- 60.0s
  0.0s-8.0s:  I want to say any more. Okay, don't say any more. Let's just listen to this dialogue and we'll be back later to explain it.
60.0s -- 90.0s
  0.0s-7.7s:  Does it come with Coke and fries? It comes with either soup or salad and a complimentary
  7.7s-17.6s:  glass of wine, sir. I'll go with the spaghetti and meatballs, salad and the wine. Excellent choice.
  17.6s-25.8s:  Your order will be ready soon. How soon is soon? 20 minutes? You know what? I'll just go grab
90.0s -- 120.0s
  0.0s-3.0s:  I'm just gonna go to my vavapia.
  5.0s-6.8s:  Oh, oh, why do I wait there?
  6.8s-8.0s:  What a waiter.
  8.0s-10.9s:  He is really angry at this customer
  10.9s-12.8s:  and I can't blame him.
  12.8s-14.4s:  I would be angry too.
  14.4s-15.4s:  Yeah.
  15.4s-19.6s:  So I think Marco, you have chosen some interesting
  19.6s-21.4s:  expressions for us here.
  21.4s-22.5s:  What is the first one?
  22.5s-26.6s:  Well, the first one is I'm still working on it.
  26.6s-28.4s:  I'm still working on it.
120.0s -- 150.0s
  0.0s-9.1s:  on it. I'm still working on it. Why don't we listen to some other examples on how you can use this word and then we'll come back and explain it.
150.0s -- 180.0s
  0.0s-28.0s:  Ahhh, so you could say I'm still working on it means I still need more time. Fantastic. Great. Okay. It's clear. I have another one for you guys. It's a phrase commonly used and you'll hear it all the time. Coke and fries. Coke and fries. Coke and fries. Coke and fries. Coke and fries. It's a short way of saying Coca-Cola and French fries.
180.0s -- 210.0s
  0.0s-29.0s:  Yeah, burger, Coke, and fries. Okay. What's next on the menu? On the menu we have a complimentary glass of wine. Complimentary glass of wine. Complimentary means that it's free. Free. Yes. We love that, huh? Yeah, we definitely like free things. Well, I have another one for us here. Another interesting phrase, I'll go with.
210.0s -- 240.0s
  0.0s-29.2s:  I'll go with. I'll go with. So basically it's another way of saying I'm choosing. Yeah, I'll choose or I'll take. I'll take. Mm-hmm. Okay. Very good. All right. The next word that I want to take a look at is grab. Grab. Grab. Grab. Okay. Now let's listen to some other examples of how we could use grab in different situations and then we'll come back and explain it.
240.0s -- 270.0s
  0.0s-29.9s:  Let's grab a cup of coffee. All right, well, I would understand that go grab means, go get quickly.
270.0s -- 300.0s
  0.0s-29.6s:  go get quickly, yeah. Yeah. All right, everyone. So now it's time to listen one more time to the dialogue and listen carefully and try to understand why the waiter doesn't like the customer. Let's listen. Good evening. My name is Fabio. I'll be your waiter for tonight. May I take your order? No, I'm still working on it. This menu is not even in English.
300.0s -- 330.0s
  0.0s-7.4s:  What's good here? For you sir, I would recommend spaghetti and meatballs. Does it come with
  7.4s-17.7s:  Coke and fries? It comes with either soup or salad and complimentary glass of wine, sir. I'll go
  17.7s-26.4s:  with the spaghetti and meatballs, salad and the wine. Excellent choice. Your order will be ready soon.
330.0s -- 360.0s
  0.0s-3.0s:  You know what? I'll just go grab a burger across the street.
```
